File: 0047-permutations-ii/0047-permutations-ii.cpp
class Solution {
public:
    vector<vector<int>> ans;
    vector<vector<int>> permuteUnique(vector<int>& nums) {
        int *hash = new int[nums.size() + 1];
        vector<int> a;
        solve(nums, a, hash, 0);
        return ans;
    }
    
    void solve(vector<int>& nums, vector<int> a, int hash[], int j){
        if( j == nums.size()){
            if (std::find(ans.begin(), ans.end(),a)==ans.end())
                 ans.push_back(a);
          return;
 }
        for(int i = 0; i < nums.size(); ++i){
            if(hash[i] == 1)
                continue;
            a.push_back(nums[i]);
            hash[i] = 1;
            solve(nums, a, hash, j + 1);
            a.pop_back();
            hash[i] = 0;
            
        }
    }
};

File: 0047-permutations-ii/0047-permutations-ii.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Solution {

    Map<Integer, Integer> frequency = new HashMap<>();
    List<List<Integer>> permutations = new ArrayList<>();

    public List<List<Integer>> permuteUnique(int[] nums) {
        for (int n : nums) {
            frequency.put(n, frequency.getOrDefault(n, 0) + 1);
        }
        findPermutations(new ArrayList<>(), nums.length);
        return permutations;
    }

    private void findPermutations(List<Integer> current, int permutationSize) {
        if (current.size() == permutationSize) {
            permutations.add(new ArrayList<>(current));
            return;
        }

        for (int n : frequency.keySet()) {

            if (frequency.get(n) > 0) {
                frequency.put(n, frequency.get(n) - 1);
                current.add(n);

                findPermutations(current, permutationSize);

                frequency.put(n, frequency.get(n) + 1);
                current.remove(current.size() - 1);
            }
        }
    }
}